form/data_mappers.rst

```diff
@@ -177,19 +181,41 @@ method to configure the data mapper::
             // when creating a new color, the initial data should be null
             $resolver->setDefault('empty_data', null);
         }
+
+        // ...
     }
 
-Cool! When using the ``ColorType`` form, the custom ``ColorMapper`` will create
-a new ``Color`` object now.
+Cool! When using the ``ColorType`` form, the custom data mapper methods will
+create a new ``Color`` object now.
 
 .. caution::
 
     When a form has the ``inherit_data`` option set to ``true``, it does not use the data mapper and
     lets its parent map inner values.
 
-.. tip::
+.. sidebar:: Stateful Data Mappers
+
+    Sometimes, data mappers need to access services or need to maintain their
```

Comment:
> Sometimes, data mappers need to access services

This is actually wrong. When a data mapper needs a service, you can still implement `DataMapperInterface` in `FormType` and add the service to the constructor of the form type.

Indeed the only legit reason for creating a separate data mapper class if you want to use form options in the data mapper.

```
$form = $this->createForm(SomeType::class, $data, ['option_for_data_mapper' => 123]);
```

Since option values are specific to `Form` instances and not to the single `FormType` instance, we need multiple data mappers with different states here. Otherwise you can implement `DataMapperInterface` in `FormType`.

Reply:
Ah, thanks for clarifying. I was already wondering about the same thing.